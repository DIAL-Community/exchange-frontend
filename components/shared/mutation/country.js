import { gql } from '@apollo/client'

export const DELETE_COUNTRY = gql`
  mutation DeleteCountry($id: ID!) {
    deleteCountry(id: $id) {
      country {
        id
        slug
        name
      }
      errors
    }
  }
`

export const CREATE_COUNTRY = gql`
  mutation CreateCountry(
    $name: String!
    $slug: String!
    $description: String
  ) {
    createCountry(
      name: $name
      slug: $slug
      description: $description
    ) {
        country {
          id
          name
          slug
          code
          codeLonger
          description
        }
        errors
      }
    }
`
