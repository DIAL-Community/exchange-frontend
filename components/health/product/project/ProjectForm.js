import React, { useState, useCallback, useContext } from 'react'
import { useRouter } from 'next/router'
import { useMutation, useQuery } from '@apollo/client'
import { useIntl } from 'react-intl'
import { FaSpinner } from 'react-icons/fa6'
import { Controller, useForm } from 'react-hook-form'
import { ToastContext } from '../../../../lib/ToastContext'
import { useUser } from '../../../../lib/hooks'
import Input from '../../../shared/form/Input'
import ValidationError from '../../../shared/form/ValidationError'
import { HtmlEditor } from '../../../shared/form/HtmlEditor'
import { CREATE_PROJECT } from '../../../shared/mutation/project'
import UrlInput from '../../../shared/form/UrlInput'
import { Error, Loading, NotFound, Unauthorized } from '../../../shared/FetchStatus'
import { DEFAULT_PAGE_SIZE } from '../../../utils/constants'
import {
  PAGINATED_PROJECTS_QUERY,
  PROJECT_PAGINATION_ATTRIBUTES_QUERY,
  PROJECT_SEARCH_QUERY
} from '../../../shared/query/project'
import Select from '../../../shared/form/Select'
import { COUNTRY_SEARCH_QUERY } from '../../../shared/query/country'

const ProjectForm = React.memo(({
  project,
  toggleCreateProjectDialog,
  projectsInput
}) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const slug = project?.slug ?? ''

  const { user, isAdminUser, isEditorUser, loadingUserSession } = useUser()

  const [mutating, setMutating] = useState(false)
  const [reverting, setReverting] = useState(false)

  const [countryValue, updateCountryValue] = useState(null)

  const { showSuccessMessage, showFailureMessage } = useContext(ToastContext)

  const router = useRouter()
  const { locale } = router

  const [updateProject, { reset }] = useMutation(CREATE_PROJECT, {
    refetchQueries: [{
      query: PROJECT_PAGINATION_ATTRIBUTES_QUERY,
      variables: { search: '' }
    }, {
      query: PAGINATED_PROJECTS_QUERY,
      variables: { search: '', limit: DEFAULT_PAGE_SIZE, offset: 0 }
    },
    {
      query: PROJECT_SEARCH_QUERY,
      variables: {
        search: projectsInput
      }
    }],
    onCompleted: (data) => {
      if (data.createProject.project && data.createProject.errors.length === 0) {
        setMutating(false)
        showSuccessMessage(format('toast.submit.success', { entity: format('ui.project.label') }))
        toggleCreateProjectDialog()
      } else {
        showFailureMessage(format('toast.submit.failure', { entity: format('ui.project.label') }))
        setMutating(false)
        reset()
      }
    },
    onError: () => {
      showFailureMessage(format('toast.submit.failure', { entity: format('ui.project.label') }))
      setMutating(false)
      reset()
    }
  })

  const {
    handleSubmit,
    register,
    control,
    formState: { errors }
  } = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    shouldUnregister: true,
    defaultValues: {
      name: project?.name,
      aliases: project?.aliases?.length ? project?.aliases.map((value) => ({ value })) : [{ value: '' }],
      projectUrl: project?.projectWebsite,
      projectCountry: project?.countries,
      description: project?.projectDescription?.description
    }
  })

  const { loading: loadingCountries, error: errorCountries , data: dataCountries } =
    useQuery(COUNTRY_SEARCH_QUERY, {
      variables: {
        search: ''
      }
    })

  const fetchedCountriesCallback = (data) => (
    data.countries?.map((country) => ({
      id: country.id,
      name: country.name,
      slug: country.slug,
      label: country.name
    }))
  )

  if (loadingCountries) {
    return <Loading />
  } else if (errorCountries) {
    return <Error />
  } else if (!dataCountries?.countries) {
    return <NotFound />
  }

  const doUpsert = async (data) => {
    if (user) {
      // Set the loading indicator.
      setMutating(true)
      // Pull all needed data from session and form.
      const { userEmail, userToken } = user
      const {
        name,
        imageFile,
        projectUrl,
        description,
        projectCountry
      } = data
      // Send graph query to the backend. Set the base variables needed to perform update.
      const variables = {
        name,
        slug,
        projectUrl,
        description,
        countrySlugs: projectCountry
      }
      if (imageFile) {
        variables.imageFile = imageFile[0]
      }

      updateProject({
        variables,
        context: {
          headers: {
            'Accept-Language': locale,
            Authorization: `${userEmail} ${userToken}`
          }
        }
      })
    }
  }

  const cancelForm = () => {
    setReverting(true)
    toggleCreateProjectDialog()
  }

  return loadingUserSession
    ? <Loading />
    : isAdminUser || isEditorUser ?
      <form onSubmit={handleSubmit(doUpsert)}>
        <div className='px-4 lg:px-0 py-4 lg:py-6 text-dial-plum'>
          <div className='flex flex-col gap-y-6 text-sm'>
            <div className='text-xl font-semibold'>
              {project
                ? format('app.editEntity', { entity: project.name })
                : `${format('app.createNew')} ${format('ui.project.label')}`}
            </div>
            <div className='flex flex-col gap-y-2'>
              <label className='required-field' htmlFor='name'>
                {format('project.name')}
              </label>
              <Input
                {...register('name', { required: format('validation.required') })}
                id='name'
                placeholder={format('project.name')}
                isInvalid={errors.name}
              />
              {errors.name && <ValidationError value={errors.name?.message} />}
            </div>
            <div className='flex flex-col gap-y-2'>
              <label htmlFor='projectUrl'>
                {format('project.url')}
              </label>
              <Controller
                id='projectUrl'
                name='projectUrl'
                control={control}
                render={({ field: { value, onChange } }) => (
                  <UrlInput value={value} onChange={onChange} id='projectUrl' placeholder={format('project.url')} />
                )}
              />
            </div>
            <div className='flex flex-col gap-y-2'>
              <label className='required-field' htmlFor='projectCountry'>
                {format('ui.country.label')}
              </label>
              <Controller
                id='projectCountry'
                name='projectCountry'
                control={control}
                render={({ field }) => (
                  <Select
                    options={fetchedCountriesCallback(dataCountries)}
                    placeholder={format('app.productStage')}
                    onChange={(value) => {
                      field.onChange(value.slug)
                      updateCountryValue(value)
                    }}
                    value={countryValue}
                  />
                )}
              />
            </div>
            <div className='flex flex-col gap-y-2'>
              <label className='required-field'>{format('project.description')}</label>
              <Controller
                name='description'
                control={control}
                render={({ field: { value, onChange } }) => (
                  <HtmlEditor
                    editorId='description-editor'
                    onChange={onChange}
                    initialContent={value}
                    placeholder={format('project.description')}
                    isInvalid={errors.description}
                  />
                )}
                rules={{ required: format('validation.required') }}
              />
              {errors.description && <ValidationError value={errors.description?.message} />}
            </div>
            <div className='flex flex-wrap text-base mt-6 gap-3'>
              <button type='submit' className='submit-button' disabled={mutating || reverting}>
                {`${format('app.submit')} ${format('ui.project.label')}`}
                {mutating && <FaSpinner className='spinner ml-3' />}
              </button>
              <button type='button' className='cancel-button' disabled={mutating || reverting} onClick={cancelForm}>
                {format('app.cancel')}
                {reverting && <FaSpinner className='spinner ml-3' />}
              </button>
            </div>
          </div>
        </div>
      </form>
      : <Unauthorized />
})

ProjectForm.displayName = 'ProjectForm'

export default ProjectForm
