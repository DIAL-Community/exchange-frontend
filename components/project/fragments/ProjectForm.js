import { memo, useCallback, useContext, useState } from 'react'
import { useRouter } from 'next/router'
import { Controller, useForm } from 'react-hook-form'
import { FaSpinner } from 'react-icons/fa6'
import { useIntl } from 'react-intl'
import { useMutation } from '@apollo/client'
import { GRAPH_QUERY_CONTEXT } from '../../../lib/apolloClient'
import { ToastContext } from '../../../lib/ToastContext'
import GeocodeAutocomplete from '../../shared/form/GeocodeAutocomplete'
import { HtmlEditor } from '../../shared/form/HtmlEditor'
import Input from '../../shared/form/Input'
import Pill from '../../shared/form/Pill'
import UrlInput from '../../shared/form/UrlInput'
import ValidationError from '../../shared/form/ValidationError'
import { CREATE_PROJECT } from '../../shared/mutation/project'
import { PAGINATED_PROJECTS_QUERY, PROJECT_PAGINATION_ATTRIBUTES_QUERY } from '../../shared/query/project'
import { DEFAULT_PAGE_SIZE } from '../../utils/constants'

const ProjectForm = memo(({ project }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const [location, setLocation] = useState()
  const [locationName, setLocationName] = useState(project.location)

  const slug = project?.slug ?? ''
  const countrySlugs = project?.countries.map(c => c.slug) ?? []

  const [mutating, setMutating] = useState(false)
  const [reverting, setReverting] = useState(false)

  const { showSuccessMessage, showFailureMessage } = useContext(ToastContext)

  const router = useRouter()
  const { locale } = router

  const LOCATION_NAME_PARTS_SEPARATOR = ', '
  const addLocation = (locationToAdd) => {
    if (locationToAdd) {
      const { cityName, regionName, countryCode, latitude, longitude } = locationToAdd
      const location = {
        name: [cityName, regionName, countryCode].join(LOCATION_NAME_PARTS_SEPARATOR),
        cityName,
        regionName,
        countryCode,
        latitude,
        longitude
      }
      setLocation(location)
      setLocationName(location.name)
    }
  }

  const removeLocation = () => {
    setLocation(null)
    setLocationName('')
  }

  const [updateProject, { reset }] = useMutation(CREATE_PROJECT, {
    refetchQueries: [{
      query: PROJECT_PAGINATION_ATTRIBUTES_QUERY,
      variables: { search: '' },
      context: {
        headers: {
          ...GRAPH_QUERY_CONTEXT.VIEWING
        }
      }
    }, {
      query: PAGINATED_PROJECTS_QUERY,
      variables: { search: '', limit: DEFAULT_PAGE_SIZE, offset: 0 },
      context: {
        headers: {
          ...GRAPH_QUERY_CONTEXT.VIEWING
        }
      }
    }],
    onCompleted: (data) => {
      if (data.createProject.project && data.createProject.errors.length === 0) {
        setMutating(false)
        const redirectPath = `/${locale}/projects/${data.createProject.project.slug}`
        const redirectHandler = () => router.push(redirectPath)
        showSuccessMessage(
          format('toast.submit.success', { entity: format('ui.project.label') }),
          redirectHandler
        )
      } else {
        showFailureMessage(format('toast.submit.failure', { entity: format('ui.project.label') }))
        setMutating(false)
        reset()
      }
    },
    onError: () => {
      showFailureMessage(format('toast.submit.failure', { entity: format('ui.project.label') }))
      setMutating(false)
      reset()
    }
  })

  const {
    handleSubmit,
    register,
    control,
    formState: { errors }
  } = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    shouldUnregister: true,
    defaultValues: {
      name: project?.name,
      aliases: project?.aliases?.length ? project?.aliases.map((value) => ({ value })) : [{ value: '' }],
      projectUrl: project?.projectWebsite,
      description: project?.projectDescription?.description
    }
  })

  const doUpsert = async (data) => {
    // Set the loading indicator.
    setMutating(true)
    // Pull all needed data from session and form.
    const {
      name,
      imageFile,
      projectUrl,
      description
    } = data
    // Send graph query to the backend. Set the base variables needed to perform update.
    const variables = {
      name,
      slug,
      projectUrl,
      description,
      location,
      countrySlugs
    }
    if (imageFile) {
      variables.imageFile = imageFile[0]
    }

    updateProject({
      variables,
      context: {
        headers: {
          'Accept-Language': locale
        }
      }
    })
  }

  const cancelForm = () => {
    setReverting(true)
    router.push(`/${locale}/projects/${slug}`)
  }

  return (
    <form onSubmit={handleSubmit(doUpsert)}>
      <div className='px-4 lg:px-0 py-4 lg:py-6 text-dial-plum'>
        <div className='flex flex-col gap-y-6 text-sm'>
          <div className='text-xl font-semibold'>
            {project
              ? format('app.editEntity', { entity: project.name })
              : `${format('app.createNew')} ${format('ui.project.label')}`}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label className='required-field' htmlFor='name'>
              {format('project.name')}
            </label>
            <Input
              {...register('name', { required: format('validation.required') })}
              id='name'
              placeholder={format('project.name')}
              isInvalid={errors.name}
            />
            {errors.name && <ValidationError value={errors.name?.message} />}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label htmlFor='projectUrl'>
              {format('project.url')}
            </label>
            <Controller
              id='projectUrl'
              name='projectUrl'
              control={control}
              render={({ field: { value, onChange } }) => (
                <UrlInput value={value} onChange={onChange} id='projectUrl' placeholder={format('project.url')} />
              )}
            />
          </div>
          <div className='flex flex-col gap-y-2'>
            <label id='description' className='required-field'>
              {format('project.description')}
            </label>
            <Controller
              name='description'
              control={control}
              render={({ field: { value, onChange } }) => (
                <HtmlEditor
                  labelledBy='description'
                  onChange={onChange}
                  initialContent={value}
                  placeholder={format('project.description')}
                />
              )}
              rules={{ required: format('validation.required') }}
            />
            {errors.description && <ValidationError value={errors.description?.message} />}
          </div>
          <div className='flex flex-col gap-y-3'>
            <label className='flex flex-col gap-y-2 mb-2'>
              {`${format('app.searchAndAssign')} ${format('project.location')}`}
              <GeocodeAutocomplete
                value={null}
                onChange={addLocation}
              />
            </label>
          </div>
          <div className='flex flex-wrap gap-3'>
            { locationName &&
              <Pill
                label={locationName}
                onRemove={() => removeLocation()}
              />
            }
          </div>
          <div className='flex flex-wrap text-base mt-6 gap-3'>
            <button type='submit' className='submit-button' disabled={mutating || reverting}>
              {`${format('app.submit')} ${format('ui.project.label')}`}
              {mutating && <FaSpinner className='spinner ml-3' />}
            </button>
            <button type='button' className='cancel-button' disabled={mutating || reverting} onClick={cancelForm}>
              {format('app.cancel')}
              {reverting && <FaSpinner className='spinner ml-3' />}
            </button>
          </div>
        </div>
      </div>
    </form>
  )
})

ProjectForm.displayName = 'ProjectForm'

export default ProjectForm
