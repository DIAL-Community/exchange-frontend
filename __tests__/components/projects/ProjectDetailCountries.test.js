import { fireEvent, screen, waitFor } from '@testing-library/react'
import ProjectDetailCountries from '../../../components/projects/ProjectDetailCountries'
import { COUNTRY_SEARCH_QUERY } from '../../../queries/country'
import { mockRouterImplementation, mockSessionImplementation, render, waitForReactSelectToLoad } from '../../test-utils'
import CustomMockedProvider, { generateMockApolloData } from '../../utils/CustomMockedProvider'
import { countries } from './data/ProjectDetailCountries'
import { project } from './data/ProjectForm'

// Mock next-router calls.
jest.mock('next/dist/client/router')
// Mock the next-auth's useSession.
jest.mock('next-auth/client')

describe('Unit test for the ProjectDetailCountries component.', () => {
  const EDIT_BUTTON_TEST_ID = 'edit-button'
  const CANCEL_BUTTON_TEST_ID = 'cancel-button'
  const COUNTRY_SEARCH_TEST_ID = 'country-search'
  const COUNTRY_SEARCH_PLACEHOLDER = 'Type to search...'
  const COUNTRY_SEARCH_OPTION_LABEL = 'Another Country'
  const PROJECT_TEST_COUNTRY_LABEL = 'Test Country'
  const PILL_TEST_ID = 'pill'
  const PILL_REMOVE_BUTTON_TEST_ID = 'remove-button'
  const mockCountries = generateMockApolloData(COUNTRY_SEARCH_QUERY, { search: '' }, null, countries)

  beforeAll(() => {
    mockRouterImplementation()
    mockSessionImplementation()
  })

  test('Should match snapshot - without edit permission.', () => {
    const { container } = render(
      <CustomMockedProvider mocks={[mockCountries]}>
        <ProjectDetailCountries
          canEdit={false}
          project={project}
        />
      </CustomMockedProvider>
    )
    expect(container).toMatchSnapshot()
  })

  test('Should match snapshot - with edit permission.', () => {
    const { container } = render(
      <CustomMockedProvider mocks={[mockCountries]}>
        <ProjectDetailCountries
          canEdit={true}
          project={project}
        />
      </CustomMockedProvider>
    )
    expect(container).toMatchSnapshot()
  })

  test('Should match snapshot - with open editable section', async () => {
    const { container, getByTestId } = render(
      <CustomMockedProvider mocks={[mockCountries]}>
        <ProjectDetailCountries
          canEdit={true}
          project={project}
        />
      </CustomMockedProvider>
    )
    fireEvent.click(getByTestId(EDIT_BUTTON_TEST_ID))
    await screen.findByText(COUNTRY_SEARCH_PLACEHOLDER)
    await waitForReactSelectToLoad(container)
    expect(container).toMatchSnapshot()
  })

  test('Should remove a pill', async () => {
    const { container, getByTestId } = render(
      <CustomMockedProvider mocks={[mockCountries]}>
        <ProjectDetailCountries
          canEdit={true}
          project={project}
        />
      </CustomMockedProvider>
    )
    fireEvent.click(getByTestId(EDIT_BUTTON_TEST_ID))
    await screen.findByText(COUNTRY_SEARCH_PLACEHOLDER)
    fireEvent.click(getByTestId(PILL_REMOVE_BUTTON_TEST_ID))
    expect(screen.queryByTestId(PILL_TEST_ID)).toBeNull()
    expect(container).toMatchSnapshot()
  })

  test('Should add a pill and revert changes on "Cancel" button click', async () => {
    const { container, getByTestId, getByText } = render(
      <CustomMockedProvider mocks={[mockCountries]}>
        <ProjectDetailCountries
          canEdit={true}
          project={project}
        />
      </CustomMockedProvider>
    )
    fireEvent.click(getByTestId(EDIT_BUTTON_TEST_ID))
    await screen.findByText(COUNTRY_SEARCH_PLACEHOLDER)

    fireEvent.keyDown(getByTestId(COUNTRY_SEARCH_TEST_ID).childNodes[1], { key: 'ArrowDown' })
    await screen.findByText(COUNTRY_SEARCH_OPTION_LABEL)
    fireEvent.click(getByText(COUNTRY_SEARCH_OPTION_LABEL))
    expect(screen.queryAllByTestId(PILL_TEST_ID)).toHaveLength(2)
    expect(container).toMatchSnapshot()

    fireEvent.click(getByTestId(CANCEL_BUTTON_TEST_ID))
    expect(screen.queryByText(PROJECT_TEST_COUNTRY_LABEL)).toBeInTheDocument()
    expect(screen.queryByText(COUNTRY_SEARCH_OPTION_LABEL)).not.toBeInTheDocument()

    fireEvent.click(getByTestId(EDIT_BUTTON_TEST_ID))
    await screen.findByText(COUNTRY_SEARCH_PLACEHOLDER)

    expect(screen.queryByText(PROJECT_TEST_COUNTRY_LABEL)).toBeInTheDocument()
    expect(screen.queryByText(COUNTRY_SEARCH_OPTION_LABEL)).not.toBeInTheDocument()
    expect(screen.queryAllByTestId(PILL_TEST_ID)).toHaveLength(1)
  })
})
