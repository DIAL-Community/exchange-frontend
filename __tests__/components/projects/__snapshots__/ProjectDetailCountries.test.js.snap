// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for the ProjectDetailCountries component. Should add a pill and revert changes on "Cancel" button click 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="mt-12"
    data-testid="editable-section"
  >
    <div
      class="flex justify-between items-center mb-3"
    >
      <div
        class="card-title text-dial-gray-dark"
      >
        Countries
      </div>
      <div
        class="gap-3"
      />
    </div>
    <div
      class="bg-edit p-6"
    >
      <p
        class="card-title text-dial-blue mb-3"
      >
        Assign
         
        Countries
      </p>
      <label
        class="flex flex-col gap-y-2 mb-2"
        data-testid="country-search"
      >
        Search and Assign Countries
        <div
          class=" css-l45kjq-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-:r2:-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class="react-select__control css-19jej77-control"
          >
            <div
              class="react-select__value-container css-3tzzl5-ValueContainer"
            >
              <div
                class="react-select__placeholder css-19jynah-placeholder"
                id="react-select-:r2:-placeholder"
              >
                Type to search...
              </div>
              <div
                class="react-select__input-container css-6j8wv5-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-:r2:-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class="react-select__input"
                  id="async-select-id-undefined"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 0; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class="react-select__indicator-separator css-1979yuk-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="react-select__indicator react-select__dropdown-indicator css-1hmmjq6-indicatorContainer"
              >
                <svg
                  class="text-2xl"
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </label>
      <div
        class="flex flex-wrap gap-3 mt-5"
      >
        <div
          class="inline-flex items-center px-2 py-1 rounded-md bg-dial-blue-light text-lg text-dial-gray-dark"
          data-testid="pill"
        >
          Test Country
          <svg
            class="ml-3 text-2xl cursor-pointer"
            data-testid="remove-button"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </div>
        <div
          class="inline-flex items-center px-2 py-1 rounded-md bg-dial-blue-light text-lg text-dial-gray-dark"
          data-testid="pill"
        >
          Another Country
          <svg
            class="ml-3 text-2xl cursor-pointer"
            data-testid="remove-button"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </div>
      </div>
      <div
        class="flex justify-end mt-8 gap-3 text-xl"
      >
        <button
          class="submit-button"
          data-testid="submit-button"
          type="submit"
        >
          Submit
        </button>
        <button
          class="cancel-button"
          data-testid="cancel-button"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the ProjectDetailCountries component. Should match snapshot - with edit permission. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="mt-12"
    data-testid="editable-section"
  >
    <div
      class="flex justify-between items-center mb-3"
    >
      <div
        class="card-title text-dial-gray-dark"
      >
        Countries
      </div>
      <div
        class="gap-3"
      >
        <button
          class="mr-2 inline-flex items-center gap-x-1.5 bg-dial-blue px-2 py-1 rounded-md text-white"
          data-testid="edit-button"
          type="button"
        >
          <img
            alt="Edit"
            height="12px"
            src="/icons/edit.svg"
            width="12px"
          />
          <span
            class="text-sm"
          >
            Edit
          </span>
        </button>
      </div>
    </div>
    <div
      class="grid grid-cols-1 md:grid-cols-2 xl:grid-cols-3"
    >
      <div
        class="border-3 border-transparent"
      >
        <div
          class="bg-white border border-dial-gray card-drop-shadow"
        >
          <div
            class="p-4 font-semibold text-button-gray"
          >
            Test Country
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the ProjectDetailCountries component. Should match snapshot - with open editable section 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="mt-12"
    data-testid="editable-section"
  >
    <div
      class="flex justify-between items-center mb-3"
    >
      <div
        class="card-title text-dial-gray-dark"
      >
        Countries
      </div>
      <div
        class="gap-3"
      />
    </div>
    <div
      class="bg-edit p-6"
    >
      <p
        class="card-title text-dial-blue mb-3"
      >
        Assign
         
        Countries
      </p>
      <label
        class="flex flex-col gap-y-2 mb-2"
        data-testid="country-search"
      >
        Search and Assign Countries
        <div
          class=" css-l45kjq-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-:r0:-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class="react-select__control css-19jej77-control"
          >
            <div
              class="react-select__value-container css-3tzzl5-ValueContainer"
            >
              <div
                class="react-select__placeholder css-19jynah-placeholder"
                id="react-select-:r0:-placeholder"
              >
                Type to search...
              </div>
              <div
                class="react-select__input-container css-6j8wv5-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-:r0:-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class="react-select__input"
                  id="async-select-id-undefined"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class="react-select__indicator-separator css-1979yuk-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="react-select__indicator react-select__dropdown-indicator css-1hmmjq6-indicatorContainer"
              >
                <svg
                  class="text-2xl"
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </label>
      <div
        class="flex flex-wrap gap-3 mt-5"
      >
        <div
          class="inline-flex items-center px-2 py-1 rounded-md bg-dial-blue-light text-lg text-dial-gray-dark"
          data-testid="pill"
        >
          Test Country
          <svg
            class="ml-3 text-2xl cursor-pointer"
            data-testid="remove-button"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </div>
      </div>
      <div
        class="flex justify-end mt-8 gap-3 text-xl"
      >
        <button
          class="submit-button"
          data-testid="submit-button"
          disabled=""
          type="submit"
        >
          Submit
        </button>
        <button
          class="cancel-button"
          data-testid="cancel-button"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the ProjectDetailCountries component. Should match snapshot - without edit permission. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="mt-12"
    data-testid="editable-section"
  >
    <div
      class="flex justify-between items-center mb-3"
    >
      <div
        class="card-title text-dial-gray-dark"
      >
        Countries
      </div>
      <div
        class="gap-3"
      />
    </div>
    <div
      class="grid grid-cols-1 md:grid-cols-2 xl:grid-cols-3"
    >
      <div
        class="border-3 border-transparent"
      >
        <div
          class="bg-white border border-dial-gray card-drop-shadow"
        >
          <div
            class="p-4 font-semibold text-button-gray"
          >
            Test Country
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the ProjectDetailCountries component. Should remove a pill 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="mt-12"
    data-testid="editable-section"
  >
    <div
      class="flex justify-between items-center mb-3"
    >
      <div
        class="card-title text-dial-gray-dark"
      >
        Countries
      </div>
      <div
        class="gap-3"
      />
    </div>
    <div
      class="bg-edit p-6"
    >
      <p
        class="card-title text-dial-blue mb-3"
      >
        Assign
         
        Countries
      </p>
      <label
        class="flex flex-col gap-y-2 mb-2"
        data-testid="country-search"
      >
        Search and Assign Countries
        <div
          class=" css-l45kjq-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-:r1:-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class="react-select__control css-19jej77-control"
          >
            <div
              class="react-select__value-container css-3tzzl5-ValueContainer"
            >
              <div
                class="react-select__placeholder css-19jynah-placeholder"
                id="react-select-:r1:-placeholder"
              >
                Type to search...
              </div>
              <div
                class="react-select__input-container css-6j8wv5-Input"
                data-value=""
              >
                <input
                  aria-autocomplete="list"
                  aria-describedby="react-select-:r1:-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class="react-select__input"
                  id="async-select-id-undefined"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class="react-select__indicator-separator css-1979yuk-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="react-select__indicator react-select__dropdown-indicator css-1hmmjq6-indicatorContainer"
              >
                <svg
                  class="text-2xl"
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21 21l-6-6m2-5a7 7 0 11-14 0 7 7 0 0114 0z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </label>
      <div
        class="flex flex-wrap gap-3 mt-5"
      />
      <div
        class="flex justify-end mt-8 gap-3 text-xl"
      >
        <button
          class="submit-button"
          data-testid="submit-button"
          type="submit"
        >
          Submit
        </button>
        <button
          class="cancel-button"
          data-testid="cancel-button"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
